tests/clients/boto3/test_boto3_s3.py

```diff
@@ -2,277 +2,392 @@
-    assert (boto_span.ec is None)
+        # Access the event system on the S3 client
+        event_system = self.s3.meta.events
 
-    assert boto_span.data['boto3']['op'] == 'download_fileobj'
-    assert boto_span.data['boto3']['ep'] == 'https://s3.amazonaws.com'
-    assert boto_span.data['boto3']['reg'] == 'us-east-1'
-    assert boto_span.data['http']['method'] == 'POST'
-    assert boto_span.data['http']['url'] == 'https://s3.amazonaws.com:443/download_fileobj'
+        request_headers = {
+                'X-Capture-This': 'this',
+                'X-Capture-That': 'that'
+            }
+        
+        # We set the custom headers in the request context instead of params 
+        # because later in the processing of the request, there is a parameter validation step, 
+        # which doesn't allow for custom arguments. 
+        def process_custom_arguments(params, context, **kwargs):
+            if "custom_request_headers" not in context:
+                context["custom_request_headers"] = request_headers
+        
+        event_system.register('before-parameter-build.s3.CreateBucket', process_custom_arguments)
+
+        with tracer.start_active_span('test'):
+            result = self.s3.create_bucket(Bucket="aws_bucket_name")
+    
+        result = self.s3.list_buckets()
+        self.assertEqual(1, len(result['Buckets']))
+        self.assertEqual(result['Buckets'][0]['Name'], 'aws_bucket_name')
+
+        spans = tracer.recorder.queued_spans()
+        self.assertEqual(2, len(spans))
+
+        filter = lambda span: span.n == "sdk"
+        test_span = get_first_span_by_filter(spans, filter)
+        self.assertTrue(test_span)
+
+        filter = lambda span: span.n == "boto3"
+        boto_span = get_first_span_by_filter(spans, filter)
+        self.assertTrue(boto_span)
+
+        self.assertEqual(boto_span.t, test_span.t)
+        self.assertEqual(boto_span.p, test_span.s)
+
+        self.assertIsNone(test_span.ec)
+        self.assertIsNone(boto_span.ec)
+
+        self.assertEqual(boto_span.data['boto3']['op'], 'CreateBucket')
+        self.assertEqual(boto_span.data['boto3']['ep'], 'https://s3.amazonaws.com')
+        self.assertEqual(boto_span.data['boto3']['reg'], 'us-east-1')
+        self.assertDictEqual(boto_span.data['boto3']['payload'], {'Bucket': 'aws_bucket_name'})
+        self.assertEqual(boto_span.data['http']['status'], 200)
+        self.assertEqual(boto_span.data['http']['method'], 'POST')
+        self.assertEqual(boto_span.data['http']['url'], 'https://s3.amazonaws.com:443/CreateBucket')
+
+        self.assertIn("X-Capture-This", boto_span.data["http"]["header"])
+        self.assertEqual("this", boto_span.data["http"]["header"]["X-Capture-This"])
+        self.assertIn("X-Capture-That", boto_span.data["http"]["header"])
+        self.assertEqual("that", boto_span.data["http"]["header"]["X-Capture-That"])
+            
+        agent.options.extra_http_headers = original_extra_http_headers
```

Comment:
Ideally this would go into the `tearDown`, because in case of any exception or assertion failure, this won't run and has the potential to impact other test cases.